### PR TITLE
[AIRFLOW-3063] - Make DockerOperator support all docker arguments

### PR DESCRIPTION
**Pull request by MerouaneBen**
A flexible implementation of dockerOperator class that takes all the
containers config arguments in json kwargs dynamically.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3063
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
The default docker compose operator is great solution to isolate and scale the execution of dag tasks, but when it comes to use it in real complex tasks, we quickly notice that we can not use all the arguments and parameters to customise the containers available in the official Docker python lib. In this PR, we provide a flexible approach that allows to provide all the possible arguments for create_container method in kwargs object, instead of having only some of them listed in the constructor of the class.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
The original unitest of docker operator lib where updated to work with the modifications implemented in the class, and some new unit tests have been added as well.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by mik-laj** [comment]
Is your change fully backwards compatible? If it's compatible then I'm very happy.
If it is not I have another question. Is it possible to provide backward compatibility?  e.g. by adding support for two ways at the same time. If this is not possible, please add a note in the file `UPDATING.md`.

I saw that the operator does not have the full documentation described, but it would be useful. Using this operator is a common scenario. Would you like to write a guide on this topic? 
Example of a guide: 
https://airflow.readthedocs.io/en/latest/howto/operator/gcp/natural_language.html

File: airflow/operators/docker_operator.py [hunk 1]
```diff
@@ -93,86 +72,128 @@ class DockerOperator(BaseOperator):
...
+                                'command': 'echo hello world!'
+                                'environment':{'SOME_ENV_VAR':'SOME_VALUE'},
+                                'working_dir':'/container/some_path'
+            }
+    :param host_config: a json that define all host config of the container
+    :type host_config: json object
+        **Use of host_config**
+        Takes a json object only, you can provide all the supported arguments in
+        create_host_config method. Check create_host_config method documentation for
+        more details: https://goo.gl/65NcDi
```
Comment (mik-laj) [inline airflow/operators/docker_operator.py:103]: Can you use the full address?

File: airflow/operators/docker_operator.py [hunk 2]
```diff
@@ -93,86 +72,128 @@ class DockerOperator(BaseOperator):
...
     :param xcom_all: Push all the stdout or just the last line.
         The default is False (last line).
     :type xcom_all: bool
     :param docker_conn_id: ID of the Airflow connection to use
     :type docker_conn_id: str
-    :param shm_size: Size of ``/dev/shm`` in bytes. The size must be
-        greater than 0. If omitted uses system default.
-    :type shm_size: int
+    :param container_config: a json that define all the config of the container .
+    :type container_config: json object
```
Comment (mik-laj) [inline airflow/operators/docker_operator.py:84]:
```suggestion
    :type container_config: dict
```

File: airflow/operators/docker_operator.py [hunk 3]
```diff
@@ -93,86 +72,128 @@ class DockerOperator(BaseOperator):
...
+        .. code-block:: python
+
+            container_config = {'image':'ubuntu:latest',
+                                'command': 'echo hello world!'
+                                'environment':{'SOME_ENV_VAR':'SOME_VALUE'},
+                                'working_dir':'/container/some_path'
+            }
+    :param host_config: a json that define all host config of the container
+    :type host_config: json object
+        **Use of host_config**
```
Comment (mik-laj) [inline airflow/operators/docker_operator.py:100]:
```suggestion
```

File: airflow/operators/docker_operator.py [hunk 4]
```diff
@@ -235,17 +244,9 @@ def execute(self, context):
             if result['StatusCode'] != 0:
                 raise AirflowException('docker container failed: ' + repr(result))
 
-            # duplicated conditional logic because of expensive operation
-            if self.do_xcom_push:
+            if self.xcom_push_flag:
```
Comment (mik-laj) [inline airflow/operators/docker_operator.py:247]: Can you keep a comment?

**Comment by ashb** [comment]
For "common" arguments it's nice to be able to support direct kwargs - can we have a hybrid approach that keeps most of the current params but allows extra ones to be specified too?

File: airflow/operators/docker_operator.py [hunk 5]
```diff
@@ -93,86 +72,128 @@ class DockerOperator(BaseOperator):
...
+
+            container_config = {'image':'ubuntu:latest',
+                                'command': 'echo hello world!'
+                                'environment':{'SOME_ENV_VAR':'SOME_VALUE'},
+                                'working_dir':'/container/some_path'
+            }
+    :param host_config: a json that define all host config of the container
+    :type host_config: json object
+        **Use of host_config**
+        Takes a json object only, you can provide all the supported arguments in
```
Comment (ashb) [inline airflow/operators/docker_operator.py:101]: JSON object is somewhat ambigious in python docs. Is this a python dict, or a JSON string?

**Comment by MerouaneBen** [comment]
Thanks @mik-laj & @ashb for the review, the version I have commited  doesn't support backwards compatibility,  but I can implement it like @ashb has suggested.
 I'll update my PR, and fix the docstring.

**Comment by OmerJog** [comment]
@MerouaneBen any progress with this PR?

**Comment by ashb** [comment]
Please re-open a new PR with back-compat.

**Pull request closed without merging**